### ISSUE TITLE
Column vectors work but row vectors fail

### ISSUE THREAD

**Issue by Krastanov** [FluxML/Zygote.jl#331]
Here are a couple of examples that I thought were very simple, but do not work:

```
gradient(x->norm(x*[1, 1]),1.23) # works fine
gradient(x->norm(x*[1 1]),1.23) # ERROR: Mutating arrays is not supported
gradient(x->sum(x*[1 1]),1.23) # ERROR: Mutating arrays is not supported
gradient(x->(x*[1 1])[1],1.23) # ERROR: Mutating arrays is not supported
```

Same behavior with complex numbers, but different error message:

```
gradient(x->norm(x*[1im, 1]),1.23) # works fine
gradient(x->norm(x*[1im 1]),1.23) # ERROR: Compiling Tuple{typeof(Base.hvcat_fill),Array{Complex{Int64},2},Tuple{Complex{Int64},Int64}}: DimensionMismatch("dimensions must match")
gradient(x->abs(sum(x*[1im 1])),1.23) # same error
gradient(x->abs((x*[1im 1])[1]),1.23) # same error
```

On Julia 1.3-rc2 and Zygote master as of Sep 18th.

**Comment by Krastanov** [comment]
In particular, `x -> (x*[1 1])[1]` raises "Mutating arrays is not supported", but I really do not see how that expression mutates arrays.

**Comment by MikeInnes** [comment]
In practice that error comes up because you're trying to differentiate code in the standard library, which does mutate arrays under the hood. The fix is just to add whatever explicit gradient we're missing so that doesn't happen. In this case it looks like array-scalar multiplication is the issue, so we probably just need to have a method to handle that.

**Comment by mcabbott** [comment]
Here’s a work-around: 
```
julia> y = [1 1]
1×2 Array{Int64,2}:
 1  1

julia> Zygote.gradient(x -> (x*y)[1], 1.23)
(1.0,)

julia> Zygote.gradient(x -> (x*hcat(1,1))[1], 1.23)
ERROR: Mutating arrays is not supported
```
I haven’t checked, but presumably nobody thought to write a gradient for this method of `hcat`, which is what `[1 1]` really calls.

**Comment by sethaxen** [comment]
This was fixed by #847 and can be closed
